### PR TITLE
Added NT Engineering 12-key Macropad

### PR DESCRIPTION
**Pull request by NickTaormina**
## Description

Added firmware for NT Engineering 12-key Macropad

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/nt_engineering/macropad/config.h
Comment (waffle87) [inline keyboards/nt_engineering/macropad/config.h:1]: This file should be removed, as this is the default behaviour.

File: keyboards/nt_engineering/macropad/info.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    "matrix_pins": {
+        "cols": ["GP7", "GP0", "GP1"],
+        "rows": ["GP27", "GP28", "GP29", "GP6"]
+    },
+    "features": {
+        "bootmagic": false,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": false,
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/info.json:17]: This is preferred to be left enabled if possible.
```suggestion
        "mousekey": true,
```

File: keyboards/nt_engineering/macropad/info.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        "nkro": true,
+        "rgblight": true
+    },
+    "rgblight": {
+        "led_count": 1,
+        "hue_steps": 10,
+        "saturation_steps": 17,
+        "brightness_steps": 17,
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/info.json:25]:
```suggestion
```

File: keyboards/nt_engineering/macropad/info.json [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            "static_gradient": true,
+            "twinkle": true
+        }
+    },
+    "ws2812": {
+        "pin": "GP12",
+        "driver": "vendor"
+    },
+    "encoder": {
+        "enabled": true,
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/info.json:45]:
```suggestion
```

File: keyboards/nt_engineering/macropad/info.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        "rows": ["GP27", "GP28", "GP29", "GP6"]
+    },
+    "features": {
+        "bootmagic": false,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": false,
+        "nkro": true,
+        "rgblight": true
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/info.json:19]:
```suggestion
        "rgblight": true,
        "encoder": true
```

File: keyboards/nt_engineering/macropad/info.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    "ws2812": {
+        "pin": "GP12",
+        "driver": "vendor"
+    },
+    "encoder": {
+        "enabled": true,
+        "rotary": [
+            {"pin_a": "GP2", "pin_b": "GP4", "resolution": 2}
+        ]
+    },
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/info.json:49]:
```suggestion
    },
    "community_layouts": ["ortho_4x3"],
```

File: keyboards/nt_engineering/macropad/keymaps/default/config.h
Comment (waffle87) [inline keyboards/nt_engineering/macropad/keymaps/default/config.h:1]: This file should be removed.

File: keyboards/nt_engineering/macropad/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2023 Nick Taormina
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include QMK_KEYBOARD_H
+
+#define _MAIN 0
+#define _FN 1
+
+#define KC_X0 LT(_FN, KC_ESC)
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {[_MAIN] = LAYOUT_ortho_4x3(KC_ENT, KC_0, KC_BSPC, KC_7, KC_8, KC_9, KC_4, KC_5, KC_6, KC_1, KC_2, KC_3), [_FN] = LAYOUT_ortho_4x3(KC_MEDIA_PREV_TRACK, KC_MEDIA_PLAY_PAUSE, KC_MEDIA_NEXT_TRACK, KC_CALCULATOR, KC_MY_COMPUTER, KC_WWW_HOME, KC_BACKSLASH, KC_PIPE, KC_TILDE, KC_UNDO, C(KC_C), C(KC_V))};
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/keymaps/default/keymap.c:10]: Also in the keymap's current state, the `_FN` layer is not accessible.
```suggestion
enum layers {
  _MAIN,
  _FN
};

const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
  [_MAIN] = LAYOUT_ortho_4x3(
    KC_ENT, KC_0, KC_BSPC, 
    KC_7,   KC_8, KC_9,
    KC_4,   KC_5, KC_6, 
    KC_1,   KC_2, KC_3
  ),
  [_FN] = LAYOUT_ortho_4x3(
    KC_MPRV, KC_MPLY, KC_MNXT, 
    KC_CALC, KC_MYCM, KC_WHOM, 
    KC_BSLS, KC_PIPE, KC_TILDE, 
    KC_UNDO, C(KC_C), C(KC_V)
  )
};
```

File: keyboards/nt_engineering/macropad/rules.mk [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Build Options
+#   change yes to no to disable
+#
+
+
+
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/rules.mk:6]:
```suggestion
# This file intentionally left blank
```

File: keyboards/nt_engineering/macropad/readme.md
Comment (waffle87) [inline keyboards/nt_engineering/macropad/readme.md:1]: Please add bootloader instructions, as per [readme template](https://github.com/qmk/qmk_firmware/blob/master/data/templates/keyboard/readme.md?plain=1).

File: keyboards/nt_engineering/macropad/macropad.c [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (!readPin(GP26)) { // if switch is pressed
+        in_extended_layers = true;
+        if (clockwise) {
+            layer_move(2); // switch to layer 2 on clockwise rotation
+        } else {
+            layer_move(3); // switch to layer 3 on counter-clockwise rotation
+        }
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/macropad.c:75]: What exactly is the purpose of this code?

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by waffle87** [comment]
I have marked a comment as unresolved that was not addressed.

**Comment by NickTaormina** [comment]
> I have marked a comment as unresolved that was not addressed.

I reformatted the file to match what you put. Pushing the rotary encoder switches between the two layers, so there is no button in the keymap to swap them

File: keyboards/nt_engineering/macropad/macropad.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+void keyboard_post_init_kb(void) {
+    // set pin based on active status
+    rgblight_enable();
+    rgblight_sethsv(18, 255, 64);
+    keyboard_post_init_user();
+}
+
+// Handles layer switching and RGB LED
+void matrix_scan_user(void) {
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/macropad.c:41]:
```suggestion
void matrix_scan_kb(void) {
```

File: keyboards/nt_engineering/macropad/macropad.c [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                rgblight_sethsv(4, 255, 64);
+                layer_switched = true;
+            } else {
+                layer_move(0); // Move back to Layer 0
+                rgblight_sethsv(40, 255, 64);
+                layer_switched = false;
+            }
+            switch_timer = timer_read32();
+        }
+    }
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/macropad.c:56]:
```suggestion
    }
    matrix_scan_user();
```

File: keyboards/nt_engineering/macropad/macropad.c [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+// Handles rotary encoder
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (clockwise) {
+        tap_code(KC_VOLU);
+    } else {
+        tap_code(KC_VOLD);
+    }
+    return true;
+}
```
Comment (waffle87) [inline keyboards/nt_engineering/macropad/macropad.c:67]: This is the default behaviour.
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**